crates/rbuilder/src/provider/ipc_state_provider.rs

```diff
@@ -0,0 +1,580 @@
+        let span = trace_span!("bytecode", id = rand::random::<u64>());
+        let _guard = span.enter();
+        trace!("bytecode:get");
+
+        let empty_hash = code_hash.is_zero() || *code_hash == KECCAK_EMPTY;
+        if empty_hash {
+            return Ok(None);
+        }
+
+        if let Some(bytecode) = self.code_cache.get(code_hash) {
+            return Ok(Some(bytecode.clone()));
+        }
+
+        let bytecode = self
+            .ipc_provider
+            .call::<_, Option<Bytes>>("rbuilder_getCodeByHash", (code_hash,))
+            .map_err(ipc_to_provider_error)?
+            .map(|b| {
+                let bytecode = Bytecode::new_raw(b);
+                self.code_cache.insert(*code_hash, bytecode.clone());
+                bytecode
+            });
+
+        trace!("bytecode: got");
+        Ok(bytecode)
+    }
+}
+
+impl BlockHashReader for IpcStateProvider {
+    /// Get the hash of the block with the given number. Returns `None` if no block with this number exists
+    /// IMPORTANT: Assumes IPC provider (node) has RPC call:"rbuilder_getBlockHash"
+    fn block_hash(&self, number: BlockNumber) -> ProviderResult<Option<B256>> {
+        let span = trace_span!("block_hash provider", id = rand::random::<u64>());
+        let _guard = span.enter();
+        trace!("block_hash: get");
+
+        if let Some(hash) = self.block_hash_cache.get(&number) {
+            return Ok(Some(*hash));
+        }
+
+        let block_hash = self
+            .ipc_provider
+            .call::<_, Option<B256>>("rbuilder_getBlockHash", (BlockNumberOrTag::Number(number),))
+            .map_err(ipc_to_provider_error)?;
+
+        if let Some(bh) = block_hash {
+            self.block_hash_cache.insert(number, bh);
+        }
+
+        trace!("block_hash provider: got");
+        Ok(block_hash)
+    }
+
+    fn canonical_hashes_range(
+        &self,
+        _start: BlockNumber,
+        _end: BlockNumber,
+    ) -> ProviderResult<Vec<B256>> {
+        unimplemented!()
+    }
+}
+
+impl AccountReader for IpcStateProvider {
+    /// Get basic account information.
+    /// IMPORTANT: Assumes IPC provider (node) has RPC call:"rbuilder_getAccount"
+    /// Returns `None` if the account doesn't exist.
+    fn basic_account(&self, address: &Address) -> ProviderResult<Option<Account>> {
+        let span = trace_span!(
+            "account",
+            id = rand::random::<u64>(),
+            address = address.to_string()
+        );
+        let _guard = span.enter();
+        trace!("account: get");
+
+        if let Some(account) = self.account_cache.get(address) {
+            return Ok(*account);
+        }
+
+        let account = self
+            .ipc_provider
+            .call::<_, Option<AccountState>>("rbuilder_getAccount", (*address, self.block_id))
+            .map_err(ipc_to_provider_error)?
+            .map(|a| Account {
+                nonce: a.nonce.try_into().unwrap(),
```

Comment:
account is fetched from external call so maybe gracefully handle the `a.nonce.try_into` with atleast an expect error message